pkg/providers/tinkerbell/validate.go

```diff
@@ -162,14 +162,82 @@ func validateK8sVersionInOSImageURLs(spec *ClusterSpec) error {
 			}
 
 			if !containsK8sVersion(url, string(version)) {
-				return fmt.Errorf("missing kube version from worker node group machine config OSImageURL: url=%v, version=%v",
-					url, version)
+				return fmt.Errorf("missing kube version from worker node group machine config OSImageURL, the image url must include %v: url=%v, version=%v",
+					permutations(string(version), []string{"-", "_", ""}), url, version)
 			}
 		}
 	}
 	return nil
 }
 
+// permutations takes a version in dot format and
+// returns a human readable string with the permutations of the version
+// using the passed in separators.
+//
+// For example, when v = 1.29 and separators = []string{"-", "_", ""}
+// the result will be "1.29, 1-29, 1_29, or 129".
+func permutations(v string, separators []string) string {
+	result := []string{v}
+	split := strings.Split(v, ".")
+	if len(split) == 1 {
+		return v
+	}
+
+	seps := remove(".", deduplicate(separators))
+	for idx, sep := range seps {
+		var elem string
+		for idx, s := range split {
+			if idx != len(split)-1 {
+				elem = elem + s + sep
+			} else {
+				elem = elem + s
+			}
+		}
+
+		if idx == len(seps)-1 {
+			result = append(result, "or")
+		} else if idx != len(seps)-2 {
+			elem = elem + ","
+		}
+		result = append(result, elem)
+	}
+
+	if len(seps) > 1 {
+		result[0] = result[0] + ","
+	}
+
+	return strings.Join(result, " ")
+}
+
+// deduplicate returns a new slice with duplicates values removed.
+func deduplicate(s []string) []string {
```

Comment:
wondering if this is needed, since we have the control over the input and this is not coming from user?

Reply:
the control we have over the input is "business" level logic, this function is the "application" level logic. Coupling those together creates for us future issues when business logic changes. I don't recommend we couple them.